src/algebra/module/graded_module.lean

```diff
@@ -0,0 +1,103 @@
+/-
+Copyright (c) 2022 Jujian Zhang. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Jujian Zhang
+-/
+
+import ring_theory.graded_algebra.basic
+import algebra.direct_sum.decomposition
+
+/-!
+# Graded Module
+
+Given an `R`-algebra `A` graded by `𝓐`, a graded `A`-module `M` is expressed as
+`direct_sum.decomposition 𝓜` and `set_like.has_graded_smul 𝓐 𝓜`.
+Then `⨁ i, 𝓜 i` is an `A`-module and is isomorphic to `M`.
+
+## Tags
+
+graded module
+-/
+
+
+open_locale direct_sum big_operators
+
+variables {ι R A M σ : Type*}
+variables [decidable_eq ι] [add_monoid ι] [comm_semiring R] [semiring A] [algebra R A]
+variables (𝓐 : ι → submodule R A)
+variables [add_comm_monoid M] [module A M]
+variables [set_like σ M] [add_submonoid_class σ M] (𝓜 : ι → σ)
+
+namespace graded_module
+
+instance graded_algebra.to_graded_module [graded_algebra 𝓐] :
+  set_like.has_graded_smul 𝓐 (λ i, (𝓐 i).to_add_submonoid) :=
+{ smul_mem := λ i j x y hi hj, set_like.graded_monoid.mul_mem hi hj }
+
+/--
+`⨁ i, 𝓜 i` is also an `A`-module, via `a • z = decompose (a • redecompose z)` where `decompose` and
+`recompose` are the cannonical homomorphism `M → ⨁ i, 𝓜 i` and `⨁ i, 𝓜 i → M`.
+-/
+def is_module [graded_algebra 𝓐] [@direct_sum.decomposition ι M σ _ _ _ _ 𝓜] :
+  module A (⨁ i, 𝓜 i) :=
+{ smul := λ a z, direct_sum.decompose_add_equiv 𝓜 (a • (direct_sum.decompose_add_equiv 𝓜).symm z),
```

Comment:
I think you can do this without needing a decomposition on M, by summing over the decomposition on A.

I would be tempted to copy the `graded_monoid.ghas_mul` class for `smul` and see where that leads. Mostly that would give you graded multiplicative actions too.

Reply:
I have added `ghas_scalar` classes

Reply:
I have changed definition of scalar multiplication to
```lean
def scalar_hom [decidable_eq ι]
   [@set_like.has_graded_scalar ι _ A _ M _ _ _ _ 𝓐 𝓜] :
   (⨁ i, 𝓐 i) →+ (⨁ i, 𝓜 i) →+ ⨁ i, 𝓜 i :=
 direct_sum.to_add_monoid $ λ i,
   add_monoid_hom.flip $ direct_sum.to_add_monoid $ λ j, add_monoid_hom.flip $
     (direct_sum.of (λ i, 𝓜 i) _).comp_hom.comp $ gscalar_hom 𝓐 𝓜
```
Is this what you have in mind @eric-wieser

Reply:
Yes, that's precisely what I had in mind